### PR TITLE
Add offline transaction signing

### LINKED ISSUE #302: Add offline transaction signing support (Infura)
### User Story
Contributors often face issues setting up Geth locally. It would be nice to support Infura as a remote Ethereum node. This would simplify local setup.

### Description
Working with Infura requires offline transaction signing, as remote Infura node does not have access to local wallet file like Geth does. This means that transactions have to be signed with user's keys before they are being sent to Ethereum node.

### Solution
Implement offline transaction signing. Connection to Infura would then be configured by enabling an option in `config.edn` file and providing Infura node URL as a value of `:eth-rpc-url` option.

### PR DESCRIPTION
**Pull request by siphiuel**
Fixes #302

This adds support for offline transaction signing. Additionally, some unused fns were removed.

Status: Finished.

**Note:** `html2png.sh` is needed because of https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2037 (Debian systems cannot run headless)

### CODE REVIEW

File: src/clj/commiteth/eth/multisig_wallet.clj [hunk 1]
```diff
@@ -191,30 +191,6 @@
...
-  "Return vector of multisig owner addresses."
-  [bounty-addr]
-  (let [bounty-contract (load-bounty-contract bounty-addr)
-        owner-addresses (-> bounty-contract
-                            (.getOwners)
-                            .get)]
-    (if owner-addresses
-      (mapv #(.toString %) (.getValue owner-addresses))
-      [])))
-
```
Comment (martinklepsch) [inline src/clj/commiteth/eth/multisig_wallet.clj]: This and a bunch of other stuff has been deleted — is it simply not needed any more at all or is this in relation to the change introduced by this PR? If it's just dead code it would be nicer to review in separate PRs but fine for now.

Comment (siphiuel) [reply src/clj/commiteth/eth/multisig_wallet.clj]: I was prompted to check those functions because they would've been impacted by the PR. Turned out they are not used anywhere.

**Comment by martinklepsch** [comment]
I think updating the documentation in the README would also be nice, i.e. answering: "How do I use this?"

**Comment by siphiuel** [comment]
Right! Added some.

File: README.md [hunk 2]
```diff
@@ -92,14 +92,22 @@ Follow the steps [here](https://developer.github.com/apps/building-github-apps/c
...
+
+In order to launch a local geth node with the bot account unlocked issue the following command:
 
 ```
 #!/bin/bash
 geth --fast --testnet --cache=1024 --datadir=$HOME/.ropsten --verbosity 4 --port 50100 --ipcpath ~/.ropsten/geth.ipc --rpc --rpcaddr 127.0.0.1 --rpcport 8545 --rpcapi db,eth,net,web3,personal --rpccorsdomain "https://wallet.ethereum.org" --unlock "0xYOUR_ADDR" --password <(echo "YOUR_PASSPHRASE")
 ```
 
+#### Remote
+Specify Infura URL in `config.edn` under `:eth-rpc-url` key, and set `:offline-signing?` to true.
```
Comment (martinklepsch) [inline README.md]: How do I get an Infura URL? Maybe link to Infura homepage as well?

File: src/clj/commiteth/eth/multisig_wallet.clj (hunk 1 above)
Comment (martinklepsch) [reply src/clj/commiteth/eth/multisig_wallet.clj]: Ok nice. Deleting code == good.

File: README.md (hunk 2 above)
Comment (siphiuel) [reply README.md]: It's a good comment. Things that might look obvious to some are very often frustratingly opaque to others. Added some more info.

File: README.md [hunk 3]
```diff
@@ -91,14 +92,22 @@ Follow the steps [here](https://developer.github.com/apps/building-github-apps/c
...
+
+In order to launch a local geth node with the bot account unlocked issue the following command:
 
 ```
 #!/bin/bash
 geth --fast --testnet --cache=1024 --datadir=$HOME/.ropsten --verbosity 4 --port 50100 --ipcpath ~/.ropsten/geth.ipc --rpc --rpcaddr 127.0.0.1 --rpcport 8545 --rpcapi db,eth,net,web3,personal --rpccorsdomain "https://wallet.ethereum.org" --unlock "0xYOUR_ADDR" --password <(echo "YOUR_PASSPHRASE")
 ```
 
+#### Remote
+Register at [Infura](https://infura.io/signup). You will receive an email with provider URLs. Paste an URL for the network you're connecting to (Mainnet/Ropsten) into `config.edn` under `:eth-rpc-url` key, and set `:offline-signing?` to true.
```
Comment (martinklepsch) [inline README.md]: Are there any potential pitfalls around connecting to mainnet? Situations where a user could lose funds or similar?

**Comment by martinklepsch** [comment]
@rcullito @omartell You've both mentioned some trouble getting Geth set up. If you feel like you could try this PR and let us know if this improves the situation. There's no bounty for testing this but asking doesn't hurt. Thanks! 🍻

File: README.md (hunk 3 above)
Comment (siphiuel) [reply README.md]: There could be I believe. Do you think reference to Mainnet should be deleted?

Comment (martinklepsch) [reply README.md]: Maybe. Or maybe have a little warning like this (including the quote thingy):

> :warning: Be careful not to use the Mainnet unless you want to test with actual ETH tokens.

"actual ETH tokens" is probably not the right way to phrase things here but I guess you get what I mean?

Comment (siphiuel) [reply README.md]: I think it's better to remove the mention of Mainnet altogether. @andytudhope - what do you think?

Comment (martinklepsch) [reply README.md]: Go for it. We can add it back later. :)

Comment (martinklepsch) [reply README.md]: Maybe also mention that Ropsten is a network explicitly for testing / development stuff.

**Comment by rcullito** [comment]
@martinklepsch @siphiuel thank you both... this pass is definitely much more accessible! (at least for a newcomer like me). I had just two remaining questions about the setup process after going through the updated README from this branch. 

1. When using infura, what should the value of `:eth-wallet-file` be? can this be left blank?

2. A couple places in the `.edn` file mentioned the bot account. Is this referring to either the github or oauth apps setup in the previous steps, or instead the bot that posts the contract addresses on the issues? If the latter, is it possible for just a contributor to know the password for this? I wasn't sure what to enter for that field.

File: src/clj/commiteth/eth/web3j.clj [hunk 4]
```diff
@@ -25,4 +29,26 @@
                       {:password password :file-path file-path})))))
 
 (defn create-web3j []
-  (Web3j/build (HttpService. (eth/eth-rpc-url))))
+  (Web3j/build (HttpService. (env :eth-rpc-url "http://localhost:8545"))))
+
+(defn get-signed-tx [gas-price gas-limit to data]
```
Comment (pablodip) [inline src/clj/commiteth/eth/web3j.clj]: How about placing this in `commiteth.eth.core`? To me it is more a feature about Ethereum, even if we use the tool `web3j` to solve it.

File: README.md [hunk 5]
```diff
@@ -91,14 +92,22 @@ Follow the steps [here](https://developer.github.com/apps/building-github-apps/c
 
 ## Running
 
-### Geth
-Launch a local geth node with the bot account unlocked:
+### Ethereum node
+There are two options for connecting to an Ethereum node: either run a local node or connect to a remote one like Infura.
```
Comment (pablodip) [inline README.md]: Geth can also be in remote, so I would probably say `either Geth or Infura`.

File: README.md [hunk 6]
```diff
@@ -91,14 +92,22 @@ Follow the steps [here](https://developer.github.com/apps/building-github-apps/c
...
+
+In order to launch a local geth node with the bot account unlocked issue the following command:
 
 ```
 #!/bin/bash
 geth --fast --testnet --cache=1024 --datadir=$HOME/.ropsten --verbosity 4 --port 50100 --ipcpath ~/.ropsten/geth.ipc --rpc --rpcaddr 127.0.0.1 --rpcport 8545 --rpcapi db,eth,net,web3,personal --rpccorsdomain "https://wallet.ethereum.org" --unlock "0xYOUR_ADDR" --password <(echo "YOUR_PASSPHRASE")
 ```
 
+#### Remote
+Register at [Infura](https://infura.io/signup). You will receive an email with provider URLs. Paste an URL for the network you're connecting to (Mainnet/Ropsten) into `config.edn` under `:eth-rpc-url` key, and set `:offline-signing` to true.
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply README.md]: Yes, this is a normal JSON RPC url.

File: src/clj/commiteth/eth/web3j.clj (hunk 4 above)
Comment (siphiuel) [reply src/clj/commiteth/eth/web3j.clj]: I'd then say we banish `web3j` ns altogether. There's not much more in there anyway. What do you think?

File: README.md (hunk 3 above)
Comment (siphiuel) [reply README.md]: Done.

File: README.md [hunk 7]
```diff
@@ -71,6 +71,7 @@ eth-account | Ethereum account ID for the bot
 eth-password | Ethereum account password for the bot
 eth-rpc-url | RPC URL to Ethereum node, e.g. Geth. Either local or remote
 eth-wallet-file | Location of wallet file. If Geth is run with the parameters as given below, it will reside under `$HOME/.ropsten/keystore`
+offline-signing | Specifies whether to sign transactions locally before sending. Default is false. Set to true when connecting to Infura
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply README.md]: Changed that description a bit.

File: README.md (hunk 5 above)
Comment (siphiuel) [reply README.md]: Changed.

**Comment by siphiuel** [comment]
@rcullito Hi! Responses to your questions below:

1. what exactly is the "bot" account referring to? 
It actually refers to both
a.) correct account, password, and wallet file - these will be used for posting bounties and transacting contracts.
b.) github username + password - these refer to the dummy GitHub user that is used to post GitHub comments for issues with a bounty.

2. is it assumed that the :server-address should be the publicly facing equivalent of localhost:3000?
Yes

Also, what errors exactly are you getting when compiling MultiSigTokenWallet? Are you on proper versions on web3j (should be 2.3.0) and solidity (should be 0.4.15)?

**Comment by siphiuel** [comment]
I think that in order to QA this task the following two env vars have to be set prior to running the commiteth java process:

File: src/clj/commiteth/eth/web3j.clj (hunk 4 above)
Comment (pablodip) [reply src/clj/commiteth/eth/web3j.clj]: Agree! :)

File: README.md [hunk 8]
```diff
@@ -71,6 +71,7 @@ eth-account | Ethereum account ID for the bot
 eth-password | Ethereum account password for the bot
 eth-rpc-url | RPC URL to Ethereum node, e.g. Geth. Either local or remote
 eth-wallet-file | Location of wallet file. If Geth is run with the parameters as given below, it will reside under `$HOME/.ropsten/keystore`
+offline-signing | Specifies whether to sign transactions locally before sending. Default is false. Set to true when connecting to e.g. Infura or any remote Geth node that does not unlock accounts
```
Comment (pablodip) [inline README.md]: Proposal: default it to true, since that way we fix a security issue that we have right now: having the node unblocked.

**Comment by rcullito** [comment]
@siphiuel thanks for addressing all of my questions. I realize was not as clear in asking them as I might have been. Namley, is the “bot” account different from (in my case) either `rcullito` or `status-open-bounty`? I wasn’t sure if it was `status-open-bounty` - how to go about getting the password or generating the wallet file for that account? And then the same question for github user? I have credentials in place, but I am still not sure what the "bot" account refers to?

web3j is at 2.3.0 although installing solidity via brew at that specific sha was failing (Undefined symbols for architecture x86_64:) so I currently have tried with both 0.4.18 and 0.4.20

The specific errors I am receiving on `lein run`are:

```
/open-bounty/src/java/commiteth/eth/contracts/MultiSigTokenWallet.java:584: error: <identifier> expected
        return executeRemoteCallSingleValueReturn(function, List<String>.class);
                                                                         ^
4 errors
Compilation of Java sources(lein javac) failed.
```

These lead me to believe I must have the wrong wallet file in my config to begin with 🤔 
Thanks again for helping me work through it. 🙏

**Comment by siphiuel** [comment]
@rcullito You're welcome!

You can use your GitHub user and your Ethereum account for the bot. There's nothing specific about it.

**Comment by siphiuel** [comment]
Don't think it is possible to use credentials in Web3j without `WalletUtils/loadCredentials`.

File: src/clj/commiteth/eth/web3j.clj (hunk 4 above)
Comment (siphiuel) [reply src/clj/commiteth/eth/web3j.clj]: Removed.

File: README.md (hunk 8 above)
Comment (siphiuel) [reply README.md]: Agree, changed to true, both there and in `eth/core.clj`.

Comment (pablodip) [reply README.md]: Could you change it in `env/dev/resources/config.edn` as well?

Comment (siphiuel) [reply README.md]: Done:)

**Comment by churik** [comment]
Already on prod env.
Created #346.

### Environments
* Ethereum wallet used: MetaMask
* Browser: Chrome
* Operating System: macOS High Sierra 10.13.3

### Regression
* Creating new issue with bounty;
* Signing up new repo for bounty issues;
* Create several claims for bounty;
* Confirming bounty payments;
* Merging PR - closing issue;

@siphiuel would be nice if you could help with actually offline transaction feature

**Comment by churik** [comment]
Because PR is already in master, merged to develop.
